src/pykeen/pipeline.py

```diff
@@ -998,12 +998,28 @@ def pipeline(  # noqa: C901
 
     # Train like Cristiano Ronaldo
     training_start_time = time.time()
-    losses = training_loop_instance.train(
-        stopper=stopper,
-        result_tracker=result_tracker,
-        clear_optimizer=clear_optimizer,
-        **training_kwargs,
-    )
+    while True:
+        try:
+            losses = training_loop_instance.train(
+                stopper=stopper,
+                result_tracker=result_tracker,
+                clear_optimizer=clear_optimizer,
+                **training_kwargs,
+            )
+        except (MemoryError, RuntimeError) as e:
+            # If the training failed due to OOM on the GPU, the training is reverted to CPU training.
+            if training_loop_instance.device.type == 'cuda':
+                logging.warning("Tried to train the current model on GPU, but the model is too big for the GPU")
+                logging.warning("Reverting to CPU now, which will increase the training time significantly.")
+                training_loop_instance.model._set_device('cpu')
```

Comment:
Would it not better to transfer responsibility of changing the device to the user since training can take a long time and the user should be aware of oit from the beginning?

Reply:
That's a bit tricky. Because first of all, we resolve the device automatically, i.e. when PyKEEN finds a GPU and the user didn't ask for a specific device, we will use the GPU.

Aside that this leaves us with two scenarios
1. The user wants to train this specific model with these exact parameters, which does not fit on the GPU. Raising and exception and crashing the pipeline is just being annoying to the user, because then the user has to start everything from the beginning in addition has to explicitly configure the pipeline to use the CPU, which effectively doesn't change anything. In this scenario this solution is the only sensible one.
2. The user is just trying out different configurations and suddenly goes into territory that is out of GPU memory bounds. In this case this solution does not simply crash and confront the user with what just happened, but rather warns the user that the current settings are out of GPU memory bounds while still empowering him/her to make the decision to keep on running that configuration or cancelling.

In my opinion, in all scenarios this new solution is much better than the previous version.